File: 백준/Silver/14465. 소가 길을 건너간 이유 5/소가 길을 건너간 이유 5.py
import sys

input = sys.stdin.readline


n, k, b = map(int, input().split())
brokenLights = [False] * (n + 1)
for _ in range(b):
    light = int(input())
    brokenLights[light] = True

start = 1
end = 1
count = int(brokenLights[1])
answer = n

while end < n + 1:
    currLength = end - start + 1
    if currLength <= k:
        if currLength == k:
            answer = min(answer, count)
        end += 1
        if end < n and brokenLights[end]:
            count += 1
    else:
        if brokenLights[start]:
            count -= 1
        start += 1

print(answer)

File: 백준/Silver/14465. 소가 길을 건너간 이유 5/소가 길을 건너간 이유 5.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.StringTokenizer;

public class Main {

  static StringBuilder sb = new StringBuilder();
  static int n, k, b;
  static boolean[] isBroken;

  public static void main(String[] args) throws IOException {
    setUp();

    int left = 1;
    int right = 1;
    int answer = Integer.MAX_VALUE;
    int count = isBroken[1] ? 1 : 0;
    while (right < n + 1) {
      int currLength = right - left + 1;
      if (currLength == k) {
        answer = Math.min(answer, count);
        right++;
        count = isBroken[right] ? count + 1 : count;
      } else if (currLength <= k) {
        right++;
        count = isBroken[right] ? count + 1 : count;
      } else {
        count = isBroken[left] ? count - 1 : count;
        left++;
      }
    }

    sb.append(answer);

    output();
  }

  private static void setUp() throws IOException {
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    StringTokenizer st = new StringTokenizer(br.readLine());
    n = Integer.parseInt(st.nextToken());
    k = Integer.parseInt(st.nextToken());
    b = Integer.parseInt(st.nextToken());
    isBroken = new boolean[n + 2];
    for (int i = 0; i < b; i++) {
      isBroken[Integer.parseInt(br.readLine())] = true;
    }
  }

  private static void output() throws IOException {
    BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    bw.write(sb.toString());
    bw.flush();
    bw.close();
  }

}